unified_planning/engines/compilers/trajectory_constraints_remover.py

```diff
@@ -0,0 +1,422 @@
+        )
+
+    def _build_constraint_list(self, expression_quantifier_remover, env, constraints):
+        C_temp = (env.expression_manager.And(constraints)).simplify()
+        C_list = C_temp.args if C_temp.is_and() else [C_temp]
+        C_to_return = (
+            env.expression_manager.And(
+                self._remove_quantifier(expression_quantifier_remover, C_list)
+            )
+        ).simplify()
+        return C_to_return.args if C_to_return.is_and() else [C_to_return]
+
+    def _remove_quantifier(self, expression_quantifier_remover, C):
+        new_C = []
+        for c in C:
+            assert c.node_type is not OperatorKind.EXISTS
+            new_C.append(
+                expression_quantifier_remover.remove_quantifiers(c, self._problem)
+            )
+        return new_C
+
+    def _manage_sa_compilation(self, env, phi, psi, m_atom, a, E):
+        R1 = (self._regression(env, phi, a)).simplify()
+        R2 = (self._regression(env, psi, a)).simplify()
+        if phi != R1 or psi != R2:
+            cond = (
+                env.expression_manager.And([R1, env.expression_manager.Not(R2)])
+            ).simplify()
+            self._add_cond_eff(env, E, cond, env.expression_manager.Not(m_atom))
+        if psi != R2:
+            self._add_cond_eff(env, E, R2, m_atom)
+
+    def _manage_sometime_compilation(self, env, phi, m_atom, a, E):
+        R = (self._regression(env, phi, a)).simplify()
+        if R != phi:
+            self._add_cond_eff(env, E, R, m_atom)
+
+    def _manage_sb_compilation(self, env, phi, psi, m_atom, a, E):
+        R_phi = (self._regression(env, phi, a)).simplify()
+        if R_phi == phi:
+            added_precond = (None, False)
+        else:
+            rho = (
+                env.expression_manager.Or([env.expression_manager.Not(R_phi), m_atom])
+            ).simplify()
+            added_precond = (rho, True)
+        R_psi = (self._regression(env, psi, a)).simplify()
+        if R_psi != psi:
+            self._add_cond_eff(env, E, R_psi, m_atom)
+        return added_precond
+
+    def _manage_amo_compilation(self, env, phi, m_atom, a, E):
+        R = (self._regression(env, phi, a)).simplify()
+        if R == phi:
+            return None, False
+        else:
+            rho = (
+                env.expression_manager.Or(
+                    [
+                        env.expression_manager.Not(R),
+                        env.expression_manager.Not(m_atom),
+                        phi,
+                    ]
+                )
+            ).simplify()
+            self._add_cond_eff(env, E, R, m_atom)
+            return rho, True
+
+    def _manage_always_compilation(self, env, phi, a):
+        R = (self._regression(env, phi, a)).simplify()
+        if R == phi:
+            return None, False
+        else:
+            return R, True
+
+    def _add_cond_eff(self, env, E, cond, eff):
```

Comment:
I think it would be a good ide to document that you cahnge `E` or make the function returning the things to add to make it reentrant